chain/network/src/accounts_data/mod.rs

```diff
@@ -62,24 +70,113 @@ pub struct CacheSnapshot {
+    ///   so that the node literally never signs AccountsData with colling versions.
+    ///   This assumption is fragile as long as validators migrate their nodes without copying over
+    ///   the whole storage.
     fn is_new(&self, d: &SignedAccountData) -> bool {
         self.keys.contains(&d.account_key)
             && match self.data.get(&d.account_key) {
-                Some(old) if old.version >= d.version => false,
+                Some(old) if (old.version, old.timestamp) >= (d.version, d.timestamp) => false,
                 _ => true,
             }
     }
 
-    fn try_insert(&mut self, d: Arc<SignedAccountData>) -> Option<Arc<SignedAccountData>> {
+    /// Inserts d into self.data, if
+    /// * `d.account_data` is in self.keys AND
+    /// * `d.version > self.data[d.account_data].version`.
+    /// If d would override local for this node, an AccountData based on `self.local` is signed
+    /// and inserted instead to rollback the overriding change (it can happen in case the node has
+    /// been restarted and we observe the old value emitted by the previous run).
+    /// It returns the newly inserted value (or None if nothing changed).
+    /// The returned value should be broadcasted to the network.
+    fn try_insert(
+        &mut self,
+        clock: &time::Clock,
+        d: Arc<SignedAccountData>,
+    ) -> Option<Arc<SignedAccountData>> {
         if !self.is_new(&d) {
             return None;
         }
+        let d = match &self.local {
+            Some(local) if d.account_key == local.signer.public_key() => Arc::new(
+                VersionedAccountData {
+                    data: local.data.as_ref().clone(),
+                    account_key: local.signer.public_key().clone(),
+                    version: d.version + 1,
+                    timestamp: clock.now_utc(),
+                }
+                .sign(local.signer.as_ref())
+                .unwrap(),
+            ),
+            _ => d,
+        };
         self.data.insert(d.account_key.clone(), d.clone());
         Some(d)
     }
+
+    /// If `self.signer` is in `self.keys` then inserts `d` signed by `self.signer` into self.data
+    /// and returns the signed value. Otherwise does nothing and returns None.
+    /// Returned value should be broadcasted to the network.
+    /// Note that a new version of data is signed and inserted, even if it
+    /// represents the same data as the previous version - this is a way of informing the nodes on
+    /// the network that the node is actually alive.
+    /// Note that it will become critical in case we make AccountData expirable at some point.
+    fn set_local(
+        &mut self,
+        clock: &time::Clock,
+        local: LocalData,
+    ) -> Option<Arc<SignedAccountData>> {
+        let account_key = local.signer.public_key();
+        let result = match self.keys.contains(&account_key) {
+            false => None,
```

Comment:
so if they keys is not there (which would mean that we're a new validator, or we didn't get any info about ourselves from the network - we don't publish the VersionedAccountData anymore?)

Reply:
yes, unless we are not a TIER1 node (we don't participate in consensus in this and next epoch) then we don't broadcast AccountData. Even if we did, other nodes would discard it as irrelevant. We do a full sync of account data with all TIER2 peers whenever the set of TIER1 nodes changes though.